storio-common/src/main/java/com/pushtorefresh/storio/TypeMappingFinder.java

```diff
@@ -0,0 +1,93 @@
+package com.pushtorefresh.storio;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Common class for search and cache indirect type mapping.
+ */
+public class TypeMappingFinder {
+
+    @Nullable
+    private final Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping;
+
+    @NonNull
+    private final Map<Class<?>, TypeMapping<?>> indirectTypesMappingCache
+            = new ConcurrentHashMap<Class<?>, TypeMapping<?>>();
+
+    public TypeMappingFinder(@Nullable final Map<Class<?>, ? extends TypeMapping<?>> directTypesMapping) {
+        this.directTypesMapping = directTypesMapping;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <T> TypeMapping<T> findTypeMapping(@NonNull final Class<T> type) {
+
+        // Search algorithm:
+        // Check current class for direct or indirect type mapping from cache.
+        // Walk through all parent classes and interfaces.
+        // If parent type has direct mapping -> we found indirect type mapping!
+        // If current parent type == Object.class or doesn't implement another interfaces ->
+        // there is no indirect type mapping.
+        // Complexity:
+        // O(n) where n is number of parent types of passed type (pretty fast).
+
+        // Stop search if root parent is Object.class and doesn't implement another interfaces.
+
+        final TypeMapping<T> directOrCachedMapping = getDirectOrCachedTypeMapping(type);
+        if (directOrCachedMapping != null) {
+            return directOrCachedMapping;
+        }
+
+        // Okay, we don't have direct type mapping.
+        // And we don't have cache for indirect type mapping.
+        // Let's find indirect type mapping and cache it!
+
+        final Class<?> parentType = type.getSuperclass();
+        if (parentType != null && parentType != Object.class) {
+            final TypeMapping<T> mappingOfParent = (TypeMapping<T>) findTypeMapping(parentType);
```

Comment:
I have fixed it. Now it checks first level parent types just for direct type mapping and then begans to search recursively.